### DISCUSSION TITLE
Having the intercept term in least squares

### DISCUSSION THREAD

**Discussion by ogencoglu** [qpsolvers/qpsolvers#273] [Q&A] [answered]
In order to have and optimize the intercept term (i.e., `b` in `ax + b`) as well in the least squares solution, does the following makes sense:

- appending a column of `1`s to the end of `R` and
- appending a column of `0`s to the end of `A`

Do you see any issues with this approach or would you achieve the same in another way?

**Comment by stephane-caron** [comment]
You can implement and test it, <s>but I don't think it will work out this way, as we would end up adding a term colinear to the vector of 1's rather than a new optimization variable $b$</s>.

**Edit:** the intercept is the same constant $b$ across lines and your proposal is correct :+1: 

Below is my proposal for another problem :sweat_smile: where I assumed we have a different $b_i$ for each line $a_i x + b_i$.

## My misunderstanding: another problem

With an intercept term, the point is to go from $\arg\min_x \| R x - s \|$ to $\arg\min_{x,b} \| (R x + b) - s\|$, where $x \in \mathbb{R}^n$ and $b \in \mathbb{R}^m$ are optimization variables, and $R$ has shape $m \times n$. There are then $n + m$ optimization variables, so if we denote by $R'$ the new matrix passed to `solve_ls`, $R'$ should have shape $m \times (n + m)$. We can take:

```math
R' = \begin{bmatrix} R & I_{m \times m} \end{bmatrix} \Longrightarrow R' \begin{bmatrix} x \\ b \end{bmatrix} = R x + b
```

With these new optimization variables, you will indeed want to pad `A` with zeros to the right (NB: more than one column) to keep the existing constraints.

**Reply by ogencoglu** [reply, marked as answer]
Let's compare `qpsolvers` to `scikit-learn` (arguably the most popular way of fitting plain linear regression).

**Unconstrained Fit - No Intercept**
```
R = np.array([[1., 2., 0., -1], [-8., 3., 2., 4.], [0., 1., 1., 2.], [5., 2., -1., 2.], [-1., 2., 4., 3.]])
s = np.array([3., 2., -3., 0., 4.])
G, h = None, None
```

_qpsolvers_:

`x_sol = solve_ls(R, s, G, h, solver="cvxopt")` returns `[-0.03690863  1.27867596  0.98638736 -1.11230125]`.

_sklearn_:

```
model = LinearRegression(fit_intercept=False)
model.fit(R, s)
print(model.coef_)
```

returns the same `[-0.03690863  1.27867596  0.98638736 -1.11230125]`

**Unconstrained Fit - With Intercept**

_qpsolvers - my approach_:

I append a new column of `1`s.

```
R_new = np.hstack((R, np.ones((R.shape[0], 1)))).astype(float)
x_sol = solve_ls(R_new, s, G, h, solver="cvxopt")
```

which returns `[ 0.33495146  4.12135922  1.39805825 -0.98058252 -6.55825243]` and the last term is the intercept. For example, when predicting a new sample of ones: 
`(np.ones((1,4)) @ x_sol[0:-1]) + x_sol[-1]` will return `-1.68446602`.

_sklearn_:

```
model = LinearRegression(fit_intercept=True)
model.fit(R, s)
print(model.coef_, model.intercept_)
```

returns `[ 0.33495146,  4.12135922,  1.39805825, -0.98058252])` and `-6.558252427184462`, respectively. Which is the same. We can verify the same prediction of a vector ones with: `np.ones((1, 4)) @ model.coef_ + model.intercept_` which returns  `-1.68446602` again.

Another way to perform inference on a new point is the sklearn api `model.predict(np.ones((1, 4)))`

In conclusion, I think adding a single column of ones to `R` achieves the same thing. Of course for constraints, what should be done to `G` and `A` are case-dependent. So my original idea of appending zeros to `A` makes sense only for my use case but not always.

Am I missing something?

**Reply by stephane-caron** [reply]
Oh yes, you're right. As you can see in my reply, I thought you were solving $\forall i, a_i x + b_i = s_i$, while what you mean (and what everybody means, I didn't check carefully) by the intercept is $\forall i, a_i x + b = s_i$. Then your proposal is correct :+1: 

I've edited my answer above. As to the other material, well, leaving it out there as I guess it makes one more example on building problems :sweat_smile:
